package/src/bin/thing-flinger.rs

```diff
@@ -0,0 +1,621 @@
+    /// we expect the existing tools to run from 'target/debug'. Additionally,
+    // you can't run `Package` until you have actually built `omicron-package`,
+    // which `BuildMinimal` does.
+    BuildMinimal,
+
+    /// Use all subcommands from omicron-package
+    #[structopt(flatten)]
+    Package(PackageSubCommand),
+
+    /// Create an overlay directory tree for each deployment server
+    ///
+    /// Each directory tree contains unique files for the given server that will
+    /// be populated in the svc/pkg dir.
+    ///
+    /// This is a separate subcommand so that we can reconstruct overlays
+    /// without rebuilding or repackaging.
+    Overlay,
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "thing-flinger")]
+struct Args {
+    /// The path to the deployment manifest TOML file
+    #[structopt(short, long, help = "Path to deployment manifest toml file")]
+    config: PathBuf,
+
+    #[structopt(subcommand)]
+    subcommand: SubCommand,
+}
+
+/// Errors which can be returned when executing subcommands
+#[derive(Error, Debug)]
+enum FlingError {
+    #[error("Servers not listed in configuration: {0:?}")]
+    InvalidServers(Vec<String>),
+
+    /// The parameter should be the name of the argument that could not be
+    /// properly converted to a string.
+    #[error("{0} is not valid UTF-8")]
+    BadString(String),
+
+    /// Failed to rsync omicron to build host
+    #[error("Failed to sync {src} with {dst}")]
+    FailedSync { src: String, dst: String },
+
+    /// The given path must be absolute
+    #[error("Path for {field} must be absolute")]
+    NotAbsolutePath { field: &'static str },
+}
+
+// TODO: run in parallel when that option is given
+fn do_exec(
+    config: &Config,
+    cmd: String,
+    servers: Option<BTreeSet<String>>,
+) -> Result<()> {
+    if let Some(ref servers) = servers {
+        validate_servers(servers, &config.servers)?;
+
+        for name in servers {
+            let server = &config.servers[name];
+            ssh_exec(&server, &cmd, false)?;
+        }
+    } else {
+        for (_, server) in config.servers.iter() {
+            ssh_exec(&server, &cmd, false)?;
+        }
+    }
+    Ok(())
+}
+
+fn do_sync(config: &Config) -> Result<()> {
+    let server =
+        config.servers.get(&config.builder.server).ok_or_else(|| {
+            FlingError::InvalidServers(vec![config.builder.server.clone()])
+        })?;
+
+    // For rsync to copy from the source appropriately we must guarantee a
+    // trailing slash.
+    let src =
+        format!("{}/", config.local_source.canonicalize()?.to_string_lossy());
+    let dst = format!(
+        "{}@{}:{}",
+        server.username,
+        server.addr,
+        config.builder.omicron_path.to_str().unwrap()
+    );
+    let mut cmd = Command::new("rsync");
+    cmd.arg("-az")
+        .arg("-e")
+        .arg("ssh")
+        .arg("--delete")
+        .arg("--progress")
+        .arg("--exclude")
```

Comment:
Not a blocker but Iit'd be nice if we could just use .gitignore for what to exclude here.